#14 Special Shift nodes (#36)
Inbibo/shift_docs: docs/reference/nodes/iterator.md

# Iteration Nodes

Iteration Nodes are used to create and define loops within a workflow. 

> ## Iterator Node
> The *Iterator* node is used to create a loop in a workflow. This node generates multiple execution subgraphs with the nodes following the *Iterator* node. Each execution of the loop uses one value from the input iterable object as if it was effectively iterating over its elements.

Reviewer: ```suggestion
## Iterator Node

This node forces the execution of the subgraph composed by the nodes following the *Iterator* node as many times as the number of elements passed to the input of the node. Each execution of the loop uses one value from the input iterable object as if it was effectively iterating over its elements.

```

![Iterator Node](../../images/nodes/iterator.png) 

>[!NOTE]
> *Iterator* Nodes can be nested, meaning that another *Iterator* node can be placed inside the loop of an *Iterator* node to achieve multiple levels of iteration.

### Inputs

- **inValues** : This plug of type [Object](../nodes/#plugs) expects an iterable Python object, such as lists, dictionaries, sets or tuples. The number of elements of the iterable object will determine the number of loops the *Iterator* node will execute.

> ### Outputs
> - **outValue**: This plug of type [Object](../nodes/#plugs) will output one value of the input iterable object corresponding to the current iteration. This means that the output of this plug will change dynamically for each loop the iteration Node executes. 

Reviewer: ```suggestion
### Outputs

- **outValue**: This plug of type [Object](../nodes/#plugs) will output one value of the input iterable object corresponding to the current iteration. This means that the output of this plug will change dynamically for each loop the iteration Node executes. 

```

## IteratorEnd Node

> The *IteratorEnd* node is used to define the end of a loop started by an *Iterator* node. This means that all nodes between an *Iterator* and a *IteratorEnd* will be part of the execution loop. It is recommended to always have an *IteratorEnd* node per *Iterator* node in a workflow.

Reviewer: ```suggestion
The *IteratorEnd* node is used to define the end of a loop started by an *Iterator* node. This means that only the nodes between the *Iterator* and the *IteratorEnd* will compose the subgraph used for the execution loop. It is recommended to always have one *IteratorEnd* node per *Iterator* node in a workflow.

```

![IteratorEnd Node](../../images/nodes/iteratorEnd.png)  

>[!NOTE]

> > It is advised to connect to the *IteratorEnd* node the output generated in the loop via a custom `through` plug. 

Author: That all the data that must be outputted from the loop should go through the IteratorEnd operator. Otherwise, the IteratorEnd node will show up as dirty at the end of the execution. What the node does in it's execution is copying the values from the inputs to the outputs

![image](https://github.com/user-attachments/assets/7b389df6-8cc4-40dc-87ae-76f81b37e907)
VS
![image](https://github.com/user-attachments/assets/56d579c1-b020-4348-bddb-8f15d6c87686)

> > It is advised to connect to the *IteratorEnd* node the output generated in the loop via a custom `through` plug. 

Reviewer 2: Wow, i didn´t know that this was supported or designed in this way.
Then I have a question, whar value does the IteratorEnds returns? the value from the last iteration?

> > It is advised to connect to the *IteratorEnd* node the output generated in the loop via a custom `through` plug. 

Reviewer: Ideally you would accumulate the results of each iteration in whatever you pass through the iteratorEnd node so yes it will contain the value set at the last iteration.

> > It is advised to connect to the *IteratorEnd* node the output generated in the loop via a custom `through` plug. 

Reviewer 2: Thx for the explanation. Then, I would suggest to specify this in the Note. I think it's usefull to clarify that if you connect plugs in the IteratorEnd and you user the outputs latter, the value get outside the loop will be the value from the last iteration. What could be used for the ListAccumulator, if we want to add that.

> ## ListAccumulator Node

Author: List accumulator uses an instance of the list instead of a copy of the list object, and sets the result to the input as well as the output.  Basically, in each iteration, the node updates its list input. 
Given that ListAppend outputs an object copy, I understand that what you had to do is update a global variable with the output of the ListAppend. That wouldn't be needed with the ListAccumulator node.

> ## ListAccumulator Node

Reviewer 2: Ok, so the difference would be this one?
![image](https://github.com/user-attachments/assets/7b987fd2-b4f7-49b8-8a5f-242419b66642)
I'm trying to understand it to know if the documentation it's ok, because I see it a littlebit difficult to imagine how to use.

> ## ListAccumulator Node

Author: You wouldn't necessarily need to connect the list variable to the list accumulator. The default value of the inList plug is already a list.

> ## ListAccumulator Node

Reviewer: Just to add, I belive ListAccumulator nodes were developed a lot before workflow variables so what you are doing here with Variable nodes was possible before only using ListAccumulators.

> ## ListAccumulator Node

Reviewer 2: Ok I see. Before you could not save values, I sse the needed of this before the variables. nothing to modify then. Thx!
